enhancements/upgrade-bundle.md

```diff
@@ -0,0 +1,211 @@
+* `[--install-mode=]` is an optional flag that specifies the
+  [OLM install mode](https://red.ht/2YK0ksD). (see run bundle)
+  * We support 3 modes:
+    * AllNamespaces
+    * OwnNamespace
+    * SingleNamespace=ns1
+      * NOTE: if ns1 is the same as the operator's namespace, a warning will be
+        printed
+  * When `--install-mode` is not supplied, the selection algorithm is as follows:
+    * If an `OperatorGroup` already exists with the correct namespacing, use it.
+    * Else, create an `OperatorGroup` with the following preferences:
+    	* If `AllNamespaces` is supported, use this mode.
+    	* Else if `OwnNamespace` is supported, use this mode.
+    	* Else:
+			* If `SingleNamespace` is not supported, return an error saying "no valid install mode is supported"
+			* Else return an error saying "--install-mode=SingleNamespace=<namespace> is required"
+
+#### Assumptions
+* OLM is already installed
+* No previous version of the operator is already running
+* An index image exists that has the from version of the operator bundle
+* The provided images (i.e. bundle and index) are accessible and pullable by kubelet 
+
+#### Example UX Scenarios
+
+Here are some example scenarios of using `run bundle-upgrade`:
+
+* `operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0`
+
+* `operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0 --channel alpha`
+
+* `operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0 --channel alpha --namespace foo`
+
+* `operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0 --channel alpha --namespace foo --install-mode SingleNamespace=bar`
+
+#### Simple Example
+
+Here's a simple example with all the required flags.
+
+```
+operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0
+```
+
+Below is the workflow when the above command is invoked:
+
+ 1. Create pod using `--index-image`, set container command (entrypoint) to:
+    ```
+    /bin/bash -c ‘ \
+      /bin/mkdir -p /database && \
+      /bin/opm registry add   -d /database/index.db -b {.BundleImage} && \
+      /bin/opm registry serve -d /database/index.db’
+    ```
+ 1. If pod fails, clean it up, and error out with message of unsupported index image. Also capture the pod's logs for debugging.
+ 1. Create GRPC CatalogSource, pointing to `<pod-name>.<namespace>:50051`
+ 1. If an OperatorGroup doesn't exist, create one and default to `AllNamespaces` as `--install-mode` was not provided
+ 1. Create Subscription
+    * Use default values for `--channel` and `--starting-csv`, and the created `CatalogSource`'s name.
```

Comment:
How did we determine the default for the `starting-csv` I think we mention above that we pull it from the bundle, can we detail the steps here if we are going to mention it?

Reply:
`starting-csv` is required, so it has to be specified. I meant to use default value for channel alone. Rephrased it